src/variableDiagnosticsProvider.ts

```diff
@@ -0,0 +1,101 @@
+'use strict';
+
+import { workspace, languages, Diagnostic, DiagnosticSeverity, DiagnosticCollection, TextDocument, Range, Position } from 'vscode';
+
+import { OnRequestVariableEvent } from "./events/requestVariableEvent";
+import { VariableProcessor } from "./variableProcessor";
+
+export class VariableDiagnosticsProvider {
+    private httpDiagnosticCollection: DiagnosticCollection;
+
+    constructor() {
+        this.httpDiagnosticCollection = languages.createDiagnosticCollection();
+
+        this.checkVariablesInAllTextDocuments();
+
+        OnRequestVariableEvent(() => this.checkVariablesInAllTextDocuments());
+    }
+
+    public dispose(): void {
+        this.httpDiagnosticCollection.clear();
+        this.httpDiagnosticCollection.dispose();
+    }
+
+    public deleteDocumentFromDiagnosticCollection(textDocument: TextDocument) {
+        this.httpDiagnosticCollection.delete(textDocument.uri);
+    }
+
+    public checkVariablesInAllTextDocuments() {
+        workspace.textDocuments.forEach(this.checkVariables, this);
+    }
+
+    public async checkVariables(document: TextDocument) {
+        if (document.languageId !== 'http') {
+            return;
+        }
+
+        let diagnostics: Diagnostic[] = [];
+
+        let vars = this.findVariables(document);
```

Comment:
How about make the `findVariables` return `Map<string, Variable>`, so that the following logic be more simple and  achieve better performance.

```js
const varNames = [...new Set(vars.keys())]
...
if (!exists) {
  const v = vars.get(name);
diagnostics.push...
}
```

Reply:
Making a Map<string, Variable> here would disallow re-use of the same variable across lines in the document. Could make it a Map<string, Variable[]> and re-update the array-value I suppose, but I'm not sure that is more performant.

Reply:
@cbrevik sorry, it should be `Map<string, Variable[]>`. The reason I think more performant is that, in your case, the `vars` will be iterated all over for each non exist variable name to find the matched ones:
```js
existArray.forEach(({ name, exists }) => {
    if (!exists) {
        vars.forEach((v) => {
            if (v.variableName === name) {
                diagnostics.push({
                    severity: DiagnosticSeverity.Error,
                    range: new Range(new Position(v.lineNumber, v.startIndex), new Position(v.lineNumber, v.endIndex)),
                    message: `${v.variableName} is not loaded in memory`,
                    source: 'http',
                    code: 0,
                });
            }
        });
    }
});
```
While what I think we can improve is that we only need to iterate the array which corresponds to the variable name:
```js
existArray.forEach(({ name, exists }) => {
    if (!exists) {
        vars.get(name).forEach((v) => {
            diagnostics.push({
                severity: DiagnosticSeverity.Error,
                range: new Range(new Position(v.lineNumber, v.startIndex), new Position(v.lineNumber, v.endIndex)),
                message: `${v.variableName} is not loaded in memory`,
                source: 'http',
                code: 0,
            });
        });
    }
});
```

Reply:
Alright, I was thinking it was possibly not more performant because of more time complexity in `findVariables`, with the need to lookup if variable already exists in Map, and then push a new variable into the Array for that Map-key. So whatever gains we have from lookup afterwards is lost in the complexity of building the Map in the first case.

Although, I haven't really done the math, and I don't think it is that complex to really matter either way. Your way looks better as well, so I'll refactor according to your suggestion.

Reply:
This has been refactored. 👍